ddl/ddl_api.go

```diff
@@ -3897,6 +3897,184 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
 	return errors.Trace(err)
 }
 
+// getReorganizedDefinitions return the definitions as they would look like after the REORGANIZE PARTITION is done.
+func getReorganizedDefinitions(pi *model.PartitionInfo, firstPartIdx, lastPartIdx int, idMap map[int]struct{}) []model.PartitionDefinition {
```

Comment:
The partition can't be the hash partition here?

Reply:
```go
switch pi.Type {
	case model.PartitionTypeRange, model.PartitionTypeList:
	default:
		return errors.Trace(dbterror.ErrUnsupportedReorganizePartition)
	}
```

Reply:
HASH based partitioning changes by ADD PARTITION and COALESCE PARTITION, which is not yet supported. But after this feature (REORGANIZE PARTITION) the underlying core is there and we can add the support.

Reply:
https://github.com/pingcap/tidb/pull/40473/files#diff-b7c54f7dad7257cbe4fa04b8d76c9bed694e2e2e7c5599f45eb2172d8d4ff8a7R3975

For range and list only